File: 0086-partition-list/0086-partition-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def partition(self, head: Optional[ListNode], x: int) -> Optional[ListNode]:
        less = ListNode()
        greater = ListNode()
        lo = less
        hi = greater
        cur = head
        while cur:
            if cur.val < x:
                lo.next = cur
                lo = lo.next
            else:
                hi.next = cur
                hi = hi.next
            cur = cur.next
        lo.next = greater.next
        hi.next = None
        return less.next

File: 0086-partition-list/0086-partition-list.ts
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     val: number
 *     next: ListNode | null
 *     constructor(val?: number, next?: ListNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.next = (next===undefined ? null : next)
 *     }
 * }
 */

function partition(head: ListNode | null, x: number): ListNode | null {
    const less = new ListNode()
    const greater = new ListNode()
    let lo = less
    let hi = greater
    let cur = head
    for (cur = head; cur; cur = cur.next) {
        if (cur.val < x) {
            lo = lo.next = cur
        } else {
            hi = hi.next = cur
        }
    }
    lo.next = greater.next
    hi.next = null
    return less.next
};